ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,385 @@
+            } else {
+                $source->{proxy} = NCM::Component::spma::yum->generate_forward_proxy_urls($source->{protocols}, @px);
+            }
+        }
+        my $rs = EDG::WP4::CCM::TextRender->new($template, $source, relpath => 'spma', log => $self);
+        if ($rs) {
+            my $fh = $rs->filewriter("$sources_dir/$source->{name}.list", log => $self);
+            $changes += $fh->close() || 0; # handle undef
+        } else {
+            return 0;
+        }
+    }
+
+    return $changes;
+}
+
+
+# Returns a set of all installed packages
+sub get_installed_pkgs
+{
+    my $self = shift;
+    $self->debug(5, 'Entered installed_pkgs()');
+
+    my $out = CAF::Process->new($CMD_DPKG_QUERY, keeps_state => 1, log => $self)->output();
+    if ($?) {
+        return;
+    }
+    # db:Status-Abbrev is three characters, we are looking for
+    #   desired action  i (but anything is fine here)
+    #   package status  i (indicating currently installed)
+    #   error flag      a single space (indicating no error condition)
+    my @pkgs = map {s/^.{3};//} grep {m/^.i /} split("\n", $out);
+
+    return Set::Scalar->new(@pkgs);
+}
+
+# Build list of packages that should be present on system from configuration tree
+sub get_package_version_arch
+{
+    my ($self, $name, $details) = @_;
+    $self->debug(5, 'Entered get_package_version_arch()');
+
+    my @versions;
+
+    if ($details) {
+        foreach my $version (keys %$details) {
+            my $params = $details->{$version};
+            if ($params->{arch}) {
+                foreach my $arch (keys %{ $params->{arch} }) {
+                    $self->debug(5, '  Adding package ', $name, ' with version ', $version, ' and architecture ', $arch, ' to list');
+                    push(@versions, sprintf('%s:%s=%s', $name, $arch, unescape($version)));
+                }
+            } else {
+                $self->debug(5, '  Adding package ', $name, ' with version ', $version, ' but without architecture to list');
+                push(@versions, sprintf('%s=%s', $name, unescape($version)));
+            }
+        }
+    } else {
+        $self->debug(5, '  Adding package ', $name, ' without version or architecture to list');
+        push(@versions, $name);
+    }
+
+    return \@versions;
+}
+
+# Apply versions from configuration tree to a bare list of packages
+sub apply_package_version_arch
+{
+    my ($self, $packagelist, $packagetree) = @_;
+
+    $self->debug(5, 'Entered apply_package_version_arch()');
+
+    my @results;
+    my @notfound;
+
+    foreach my $name (@$packagelist) {
+        my $name_escaped = escape($name);
+        if (exists($packagetree->{$name_escaped})) {
+            my $versions = $self->get_package_version_arch($name, $packagetree->{$name_escaped});
+            push(@results, @$versions);
+        } else {
+            push(@notfound, $name);
+        }
+        $self->error("Could not find packages", join(", ", @notfound)) if @notfound;
```

Comment:
also add a space after `packages` (all arguments to reporter methods are appended, no spacing is added)